### DISCUSSION TITLE
[appDir] How to set a cookie when visiting a page

### DISCUSSION THREAD

**Discussion by Yeregorix** [vercel/next.js#46255] [Help] [answered]
I'm using NextJS 13 app directory and I want to set a cookie when an user visits a specific page.

For context, I want to implement a double submit cookie pattern to protect my forms against crsf attacks. To do that I need to set a cookie when returning the page. Of course I could use a different crsf protection strategy but I will probably have to use cookies someday for other reasons anyway.

I've searched in the docs and on github discussions but I haven't found any solution. I know that the `cookies()` function is readonly but there is no explanation in the docs as to why we are not allowed to set cookies. 

A workaround would probably be to return a client component which will then fetch an API route but this extra roundtrip could be avoided. Also this would not be javascript-disabled browser friendly.

**Comment by philwolstenholme** [comment]
I'm in a similar situation, right now I'm using a client component to do the setting.

https://beta.nextjs.org/docs/api-reference/cookies#good-to-know says:

> The Next.js team at Vercel is working on adding the ability to set cookies in addition to the cookies function.

So I think we just have to wait or use the `pages` directory/router.

**Reply by themagickoala** [reply]
This good-to-know has been updated - you can now use set in server actions and route handlers. However, this doesn't help us with app router pages!

**Comment by moshest** [comment]
You can use this package: [`next-client-cookies`](https://github.com/moshest/next-client-cookies) which will work both on client or on server (via SSR).

```ts
'use client';

import { useCookies } from 'next-client-cookies';

const MyComponent = () => {
  const cookies = useCookies();

  return (
    <div>
      <p>My cookie value: {cookies.get('my-cookie')}</p>

      <button onClick={() => cookies.set('my-cookie', 'my-value')}>
        Set cookie
      </button>
      {' | '}
      <button onClick={() => cookies.delete('my-cookie')}>Delete cookie</button>
    </div>
  );
};
```

**Reply by themagickoala** [reply]
Except you can't set cookies on the server, as your implementation uses `cookies()` under the hood, no? This doesn't help with OAuth token rotation at all.

**Reply by moshest** [reply]
I have a solution for that, it dehydrate the changes to the client once the page load.

Please note that SSR is not a server component, so I'm basically handling all 3 configurations here (SSR, client and server).

**Reply by AmirL** [reply]
So it is not possible to set HttpOnly\Secure cookies within the AppRouter?

**Reply by moshest** [reply]
Why not?

**Reply by AmirL** [reply]
Because it uses js-cookies under the hood, which works on the client side, and you can't set HttpOnly cookies on the client side.

**Reply by moshest** [reply]
I guess you're right.

**Comment by HironTez** [comment]
You can't set cookies in a server component because they start streaming the response during rendering, and it's not possible to change http headers after streaming has started. However, a possible workaround is to use middleware. You can set cookies there. But the middleware works in Edge Runtime, and if you need to perform complex actions (like accessing the DB), you'll need to fetch an API route from there and forward cookies. Here's my example of how I update tokens on the initial call:
```ts
import { parseSetCookie } from "next/dist/compiled/@edge-runtime/cookies";
import { NextRequest, NextResponse } from "next/server";
import { splitCookiesString } from "set-cookie-parser";

export async function middleware(request: NextRequest) {
  const headers = new Headers(request.headers);

  const response = NextResponse.next({
    request: {
      headers,
    },
  });

  if (request.method === "GET") {
    const responseRenewSession = await fetch(
      new URL("/renew-session", request.url),
      {
        headers,
      },
    );

    const setCookieHeader = responseRenewSession.headers.get("Set-Cookie");
    const cookieStrings = setCookieHeader
      ? splitCookiesString(setCookieHeader)
      : [];
    for (const cookieString of cookieStrings) {
      const parsed = parseSetCookie(cookieString);
      if (parsed) response.cookies.set(parsed.name, parsed.value, parsed);
    }
  }

  return response;
}
```

**Reply by Mecanik** [reply]
Making another request inside the middleware? I'm not entirely sure how optimal is that. Also, from the looks of it, `/renew-session` is already public (which is bad) and you could just do a client fetch.

**Reply by HironTez** [reply]
Yes, this workaround is not a good solution. But there's no other way to set cookies on initial request. If only they added something like 'beforeStreaming'

**Comment by Arunjenson** [comment]
any update on this?
can we set the cookie from server using route handler and able to fetch the cookie via a fetch to the route handler
is so can you provide an example

**Comment by TivonJJ** [comment]
Focus on this, I think this is really necessary！
I set the cookies via middleware now,But i think it's fussy😒

**Comment by adamjdavies** [comment]
Another alternative is to create a route handler, redirect the user to that route handler, and then redirect them to the final page

e.g. 

```typescript
redirect(`${process.env.HOST}/api/auth/register?${queryParams.toString()}`);`
``` 

and then within the route handler, set the cookie and finally redirect to your desired page:

```typescript
import { NextRequest, NextResponse } from 'next/server';
import { cookies } from 'next/headers';

export async function GET(request: NextRequest) {
 const searchParams = request.nextUrl.searchParams;
 const token = searchParams.get('token');

 if (!token) {
   return NextResponse.json({ error: 'Token is missing' }, { status: 400 });
 }

 cookies().set({
   name: 'SOME-COOKIE',
   value: token,
   httpOnly: true,
   secure: true,
   path: '/',
 });

 const urlHost = process.env.HOST || "localhost:3000";
 let redirectUrl = `${urlHost}/some-page`;

 return NextResponse.redirect(redirectUrl);
}
``` 

Hope that helps!

**Reply by devodii** [reply]
> Another alternative is to create a route handler, redirect the user to that route handler, and then redirect them to the final page
> 
> e.g.
> 
> ```ts
> redirect(`${process.env.HOST}/api/auth/register?${queryParams.toString()}`);`
> ```
> 
> and then within the route handler, set the cookie and finally redirect to your desired page:
> 
> ```ts
> import { NextRequest, NextResponse } from 'next/server';
> import { cookies } from 'next/headers';
> 
> export async function GET(request: NextRequest) {
>  const searchParams = request.nextUrl.searchParams;
>  const token = searchParams.get('token');
> 
>  if (!token) {
>    return NextResponse.json({ error: 'Token is missing' }, { status: 400 });
>  }
> 
>  cookies().set({
>    name: 'SOME-COOKIE',
>    value: token,
>    httpOnly: true,
>    secure: true,
>    path: '/',
>  });
> 
>  const urlHost = process.env.HOST || "localhost:3000";
>  let redirectUrl = `${urlHost}/some-page`;
> 
>  return NextResponse.redirect(redirectUrl);
> }
> ```
> 
> Hope that helps!

doesn't this create an endless loop of redirects between them? 🤔

**Reply by devodii** [reply]
> Another alternative is to create a route handler, redirect the user to that route handler, and then redirect them to the final page
> 
> e.g.
> 
> ```ts
> redirect(`${process.env.HOST}/api/auth/register?${queryParams.toString()}`);`
> ```
> 
> and then within the route handler, set the cookie and finally redirect to your desired page:
> 
> ```ts
> import { NextRequest, NextResponse } from 'next/server';
> import { cookies } from 'next/headers';
> 
> export async function GET(request: NextRequest) {
>  const searchParams = request.nextUrl.searchParams;
>  const token = searchParams.get('token');
> 
>  if (!token) {
>    return NextResponse.json({ error: 'Token is missing' }, { status: 400 });
>  }
> 
>  cookies().set({
>    name: 'SOME-COOKIE',
>    value: token,
>    httpOnly: true,
>    secure: true,
>    path: '/',
>  });
> 
>  const urlHost = process.env.HOST || "localhost:3000";
>  let redirectUrl = `${urlHost}/some-page`;
> 
>  return NextResponse.redirect(redirectUrl);
> }
> ```
> 
> Hope that helps!

doesn't this create an endless loop of redirects between them? 🤔

**Reply by adamjdavies** [reply]
Not necessarily. There are a couple of different things you could do to prevent this, e.g.:

- Direct the user to a different page
- Add in some logic to check if the cookie exists. If it doesn't, push them to the API route. If it does, let the page load.

**Reply by kimgysen** [reply]
This changes the browser history. That doesn't feel right to me; easy to introduce bugs that will create strange behaviour.

**Comment by AmirL** [comment]
I solve it by using middleware. 
1. Set cookie: 
```ts
const response = NextResponse.next();
response.cookies.set(...
```

 But this cookie will not be available until the next request. To use this value in the app immediately, we can set headers:
2. 
```ts
response.headers.set(...)
```

**Reply by HironTez** [reply]
You might want to set the cookie to request as well as to response

request.cookies.set(cookie);
response.cookies.set(cookie);

**Reply by AmirL** [reply]
I tried, probably it didn't work at that moment. Not sure if these cookies are available now, immediately after the middleware.

**Reply by HironTez** [reply]
That's exactly what I'm doing in my project and cookies are available in the same request

**Reply by Shawns2759** [reply]
For me the cookies are available in the next request but not the first. I ended up moving my auth to the `Authorization` header. This is updated and readable immediately.

**Comment by yiminghe** [comment]
custom middleware merchanism and cookie solution: https://github.com/yiminghe/next-compose-middlewares

**Comment by BuffaloBuffalo** [comment]
This is something we also need / find lacking with the current NextJS implementation.  `middleware` feels like it **could** be the correct option here, but it's restrictions are pretty limiting.

- You can currently only have a single middleware function defined for all routes. This means if you have middleware-esque business logic that needs to handle a specific path, we need to write a bunch of URL matching logic to apply only in specific cases.  First class support for composable middleware functions and page level middleware definitions would alleviate this.
- Edge only runtime restrictions-  Since middleware only supports edge runtime options, even if not deployed to an edge runtime. 
- Vercel's Conformance tool also flags fetch requests in middleware (`NO_FETCH_FROM_MIDDLEWARE`) which kind of seems to imply that middleware is not appropriate for this type of processing.


It kind of seems like NextJS needs a separate mechanism for this type of data fetching at the page level level for App Router. I suspect it could be a solution similar to `generateMetadata` that runs prior to headers being sent / response rendering beginning.

**Reply by yiminghe** [reply]
https://github.com/yiminghe/next-compose-middlewares can work around these problems

**Reply by mpiorowski** [reply]
The thing that you need an external lib to do sth so trivial.....that's not a good sign.

**Reply by AmirL** [reply]
It doesn't need to. There are much simpler examples above the message.

**Comment by wyattjoh** [comment, marked as answer]
You can actually set cookies via a server action: https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#cookies

This can let you re-set cookies upon form submissions, rendering the previous token useless. This doesn't however solve the initial cookie setting on navigation. This unfortunately would require the middleware to set it, as it's the only way to set a header on a GET request to an app `page.tsx`. Subsequent cookies could be re-issued by the middleware, and can be read on the page via `cookies().get()`.

The reason that the cookies can't be set during the app page render is that the streaming for it would have already started by the time that the `cookies().set()` call could have been made. This is where the server action comes into play. This would enable JS free cookie handling, as the cookie reading would be done server side, and the value could be injected into the form.

Let me know if this resolves y'alls question!

**Reply by b3arbyt3** [reply]
So, if someone uses gRPC for getting jwt or anything related, it will not work, cos gRPC is not supported using middleware? He's done?

**Comment by antoniopresto** [comment]
Sunday 01:03AM 
It really was a bad idea to go back to using nextjs 😓

**Comment by devodii** [comment]
You can read cookies from layouts and pages, but you cannot write to them.

One workaround for this is modifying the cookie inside a useEffect in a client component using the "use client" directive.

For example:
```
"use client"

export const Client = () => {
  React.useEffect(() => {
    const main = async () =>  await setCookie([{name: "key", value: value}])
    main()
  }, [])
  ```
  
  and import in the page or layout.
  
  It's also important to note that setCokie has to be exported as a server action, 
  
  i.e // cookies.ts
  
  ```
  "use server"
  
  export const setCookie = async (
  params: { name: string; value: string }[]) => {
  return params.forEach(_ => cookies().set(_.name, _.value))
}

Please let me know if this helps.

**Reply by HironTez** [reply]
If you want to set cookies on the client, just do it directly on the client. Why make an HTTP request to the server just to have it set the cookie for you?

**Reply by AmirL** [reply]
Because you can’t set secure cookies on the client side.

**Reply by HironTez** [reply]
> Because you can’t set secure cookies on the client side.

You can. There is document.cookie for this, and several libraries that help to handle it, like [js-cookie](https://www.npmjs.com/package/js-cookie).

**Reply by AmirL** [reply]
No, you can’t. Browsers don’t allow writing secure cookies on the client side.

**Reply by AmirL** [reply]
A little bit wrong, not secure, but “HttpOnly” cookies. I just use it together with Secure option.

**Reply by HironTez** [reply]
No one said anything about HttpOnly cookies. If that's what he's trying to do, opening an endpoint that sets HttpOnly cookies makes them meaningless.
